internal/pkg/service/stream/storage/node/coordinator/filerotation/conditions.go

```diff
@@ -52,18 +57,25 @@ func (c fileRotationConditionResult) String() string {
 		return "size"
 	case timeThreshold:
 		return "time"
+	case sinkThrottled:
+		return "sinkThrottled"
 	default:
 		return "unknown"
 	}
 }
 
-func shouldImport(cfg targetConfig.ImportConfig, now, openedAt, expiration time.Time, stats statistics.Value) fileRotationConditionResult {
+func shouldImport(cfg targetConfig.ImportConfig, now, openedAt, expiration time.Time, stats statistics.Value, jobLimit int) fileRotationConditionResult {
 	sinceOpened := now.Sub(openedAt).Truncate(time.Second)
 	if threshold := cfg.MinInterval.Duration(); sinceOpened < threshold {
 		// Min interval settings take precedence over other settings.
 		return newConditionResult(noConditionMet, "min interval between imports is not met")
 	}
 
+	if threshold := cfg.JobLimit; jobLimit >= threshold {
```

Comment:
My current thought process is that instead of this we could add a new plugin similar to `UploadSlice` and `ImportFile`. Something like `IsThrottling` which would let the plugin say whether it's ready to accept a new upload / import. Could even be separate `CanAcceptNewUpload` and `CanAcceptNewImport` and let the bridge make that decision. Keboola doesn't have a limit on uploads I think but a different plugin might so Keboola bridge would always return true for `CanAcceptNewUpload` and `CanAcceptNewImport` would be decided based on the jobs.

If we let the bridge decide then another question is how should it make the decision. Should we track the current jobs as you implemented it? Or should we make an api call to keboola to ask for currently running jobs and let the bridge return false if keboola is busy? (Result should be cached for a configurable amount of time so that the plugin can give the same answer if asked repeatedly in a short amount of time about the same sink.)

What do you think? And probably a better quesiton, what would Michal think? :joy:

Reply:
The idea seems leggit, as it would be plugin. But basically we always evaluate plugins, so we need to enforce each provider to implement the plugin which returns by default `false` which would not throttle or set it as default value, when not implemented.

The problem lays in effective implementation. It is possible to move the implementation of storage job under keboola bridge and use mirroring. But, I am not sure whether it is correct step to have mirror in `bridge` component.
I would not want to use caching of HTTP requests and rather use ETCD as source of truth as it can go from multiple pod's and every pod would have different cache entry etc, so it would not be cached as someone would expect.

Reply:
61160429a36a923505b729900ad9094c072ebc7e Moved the repository under `bridge`. Remove duplicate `job` model. Mirror is on top of bridge. I made it configurable in another commit. It has exported method to start mirroring as separate process. It is called under `coordinator/filerotation` module.
Created new plugin and attach it into bridge and evaluate it in `coordinator/filerotation`.
I hope it is now better as
* `+` I removed the second model, which fits more only `sink/type/tablesink/keboola` provider.
* `+` Other providers can register the plugin and evaluate, by default it is not throttled (`false`)
* `-` Bridge is now more complicated and contains repository, mirror and seems that it is more like component than library with plugins because of mirroring